types/index.d.ts

```diff
@@ -165,121 +165,159 @@ declare namespace math {
   }
 
-  interface BlockNode extends MathNode {
+  interface BlockNode<TNode extends MathNode = MathNode> extends MathNode {
     type: 'BlockNode'
     isBlockNode: true
-    blocks: Array<{ node: MathNode; visible: boolean }>
+    blocks: Array<{ node: TNode; visible: boolean }>
   }
   interface BlockNodeCtor {
-    new (
-      arr: Array<{ node: MathNode } | { node: MathNode; visible: boolean }>
+    new <TNode extends MathNode = MathNode>(
+      arr: Array<{ node: TNode } | { node: TNode; visible: boolean }>
     ): BlockNode
   }
 
-  interface ConditionalNode extends MathNode {
+  interface ConditionalNode<
+    TCond extends MathNode = MathNode,
+    TTrueNode extends MathNode = MathNode,
+    TFalseNode extends MathNode = MathNode
+  > extends MathNode {
     type: 'ConditionalNode'
     isConditionalNode: boolean
-    condition: MathNode
-    trueExpr: MathNode
-    falseExpr: MathNode
+    condition: TCond
+    trueExpr: TTrueNode
+    falseExpr: TFalseNode
   }
   interface ConditionalNodeCtor {
-    new (
-      condition: MathNode,
-      trueExpr: MathNode,
-      falseExpr: MathNode
+    new <
+      TCond extends MathNode = MathNode,
+      TTrueNode extends MathNode = MathNode,
+      TFalseNode extends MathNode = MathNode
+    >(
+      condition: TCond,
+      trueExpr: TTrueNode,
+      falseExpr: TFalseNode
     ): ConditionalNode
   }
 
-  interface ConstantNode extends MathNode {
+  interface ConstantNode<TValue extends string | number = number>
+    extends MathNode {
     type: 'ConstantNode'
     isConstantNode: true
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    value: any
+    value: TValue
   }
 
   interface ConstantNodeCtor {
-    new (constant: number): ConstantNode
+    new <TValue extends string | number = string>(
+      value: TValue
+    ): ConstantNode<TValue>
   }
 
-  interface FunctionAssignmentNode extends MathNode {
+  interface FunctionAssignmentNode<TExpr extends MathNode = MathNode>
+    extends MathNode {
     type: 'FunctionAssignmentNode'
     isFunctionAssignmentNode: true
     name: string
     params: string[]
-    expr: MathNode
+    expr: TExpr
   }
   interface FunctionAssignmentNodeCtor {
-    new (name: string, params: string[], expr: MathNode): FunctionAssignmentNode
+    new <TExpr extends MathNode = MathNode>(
+      name: string,
+      params: string[],
+      expr: TExpr
+    ): FunctionAssignmentNode<TExpr>
   }
 
-  interface FunctionNode extends MathNode {
+  interface FunctionNode<
+    TFn = SymbolNode,
+    TArgs extends MathNode[] = MathNode[]
+  > extends MathNode {
     type: 'FunctionNode'
     isFunctionNode: true
-    fn: SymbolNode
-    args: MathNode[]
+    fn: TFn
+    args: TArgs
   }
   interface FunctionNodeCtor {
-    new (fn: MathNode | string, args: MathNode[]): FunctionNode
+    new <TFn = SymbolNode, TArgs extends MathNode[] = MathNode[]>(
+      fn: TFn,
+      args: SymbolNode
+    ): FunctionNode<TransferFunction, TArgs>
```

Comment:
I don't understand why TransferFunction is used here instead of TFn.
It is not defined elsewhere in mathjs and comes from node-module/typescript/lib/lib.dom.d.ts with "version": "4.8.4" (this is the only reason why VS Code does not complain).
But when I use the mathjs 11.3.2 library in my application, the build fails as TransferFunction is undefined (my version of the typescript package is 4.3.5 and does not define type TransferFunction).

I also do not understand as well why args : SymbolNode instead of TArgs or MathNode[].

DISCLAIMER : I am not an expert of mathjs. So I may have totally missed the point.

Reply:
Oh man, that was definitely not intentional. I guess that was an accidental
VSCode autocomplete. Can you submit a PR to fix that [@dsteve](https://github.com/dsteve)?

Reply:
@mattvague so the right fix here is replace `TransferFunction` with `TFn`, right?

Reply:
I suggest this change below :
I think that args should also be changed from SymbolNode to TArgs (it looks like another VS Code autocompletion)
Do you confirm @mattvague ? (Git Copilot seems to agree as well :-)

```
  interface FunctionNodeCtor {
    new <TFn = SymbolNode, TArgs extends MathNode[] = MathNode[]>(
      fn: TFn,
      args: TArgs
    ): FunctionNode<TFn, TArgs>
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    onUndefinedFunction: (name: string) => any
  }
```

Reply:
Yes, I believe that's right. Can't believe I didn't notice this, must have been tired 🙁

Reply:
@dsteve If you don't have the time today, I can fix this. Let me know

Reply:
Thank you for your feedback, Matt. I tried to push the fix but faced some authorization failure. 
Could you please fix it, @mattvague ?

Reply:
Thanks for the proposed fix @dsteve , I'll commit this fix myself right now.

> Yes, I believe that's right. Can't believe I didn't notice this, must have been tired 🙁

No problem Matt, things can happen. I didn't spot it myself either when reviewing.

Reply:
This issue has been fixed now in v11.3.3 via 0a810abba3920411e68ffa76d4750b37fd903872.